text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,511 @@
+Spin locks are a sharp tool, which should only be used in specific circumstances (namely, when you have full control over thread scheduling).
+`#[no_std]` code might end up in user space applications with preemptive scheduling, where unbounded spin locks are inappropriate.
+
+A spin-lock based implementation of `OnceCell` is provided on crates.io in [`conquer-once`] crate.
+
+## Poisoning
+
+As a cell can be empty or fully initialized, the proposed API does not use poisoning.
+If an initialization function panics, the cell remains uninitialized.
+An alternative would be to add poisoning, which will make all subsequent `get` calls to panic.
+
+Similarly, because `OnceCell` provides strong exception safety guarantee, it implements `UnwindSafe`:
+
+```rust
+impl<T: UnwindSafe>                    UnwindSafe for OnceCell<T> {}
+impl<T: UnwindSafe + RefUnwindSafe> RefUnwindSafe for OnceCell<T> {}
+```
+
+## Default type parameter on `Lazy`
+
+`Lazy` is defined with default type parameter.
+
+```rust
+pub struct Lazy<T, F = fn() -> T> { ... }
+```
+
+This is important to make using `Lazy` in static contexts convenient.
+Without this default, the user would have to type `T` type twice:
+
+```rust
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>, fn() -> Mutex<HashMap<i32, String>>
+    = Lazy::new(|| ... );
+```
+
+If we allow type inference in statics, this could be shortened to
+
+```rust
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>, _>
+    = Lazy::new(|| ... );
+```
+
+There are two drawbacks of using fn pointer type:
+
+* fn pointers are not ZSTs, so we waste one pointer per static lazy value.
+  Lazy locals will generally rely on type-inference and will use more specific closure type.
+* Specifying type for local lazy value might be tricky: `let x: Lazy<i32> = Lazy::new(|| closed_over_var)` fails with type error, the correct syntax is `let x: Lazy<i32, _> = Lazy::new(|| closed_over_var)`.
+
+## Only thread-safe flavor
+
+It is possible to add only `sync` version of the types, as they are the most useful.
+However, this would be against zero cost abstractions spirit.
+Additionally, non thread-safe version is required to replace `thread_local!` macro without imposing synchronization.
+
+## Synchronization Guarantees
+
+In theory, it is possible to specify two different synchronization guarantees for `get` operation, release/acquire or release/consume.
+They differ in how they treat side effects.
+If thread **A** executes `get_or_init(f)`, and thread **B** executes `get` and observes the value, release/acquire guarantees that **B** also observes side-effects of `f`.
+
+Here's a program which allows to observe the difference:
+
+```rust
+static FLAG: AtomicBool = AtomicBool::new(false);
+static CELL: OnceCell<()> = OnceCell::new();
+
+// thread1
+CELL.get_or_init(|| FLAG.store(true, Relaxed));
+
+// thread2
+if CELL.get().is_some() {
+  assert!(FLAG.load(Relaxed))
+}
+```
```

Comment:
The use of consume memory order exposes the most subtle and non-intuitive properties of the memory model to the user code: happens before relationship is not transitive.

Using consume memory order in a language that doesn't support it, without explicit opt-int, in place where it will inevitably introduce subtle bugs into user code, code that would be otherwise correct, doesn't seem like a good trade-off to make for claimed performance benefits.